File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    private int f(int[] nums, int target, boolean flag){
        int low = 0, high = nums.length - 1, ans = -1;
        while(low <= high){
            int mid = low + (high - low) / 2;
            if(nums[mid] == target){
                ans = mid;
                if(flag){
                    high = mid - 1;
                } else{
                    low = mid + 1;
                }
            } else if(nums[mid] > target){
                high = mid - 1;
            } else{
                low = mid + 1;
            }
        }
        return ans;
    }
    public int[] searchRange(int[] nums, int target) {
        int ans[] = {f(nums, target, true), f(nums, target, false)};
        return ans;
    }
}

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
private:
    int f(vector<int>& nums, int target, bool flag){
        int n = nums.size(), l = 0, r = n - 1, ans = -1;
        while(l <= r){
            int m = (l + r)/2;
            if(nums[m] > target){
                r = m - 1;
            } else if(nums[m] < target){
                l = m + 1;
            } else if(flag){
                r = m - 1;  ans = m;
            } else{ 
                l = m + 1;  ans = m;
            }
        }

        return ans;
    }

public:
    vector<int> searchRange(vector<int>& nums, int target) {
        return {f(nums, target, true), f(nums, target, false)};
    }
};